### PR TITLE
Porting DTNNEmbedding Layer

### PR DESCRIPTION
**Pull request by GreatRSingh**
## Description
This PR ports the DTNNEmbedding Layer.

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [x] Documentations (modification for documents)

## Checklist

- [x] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [x] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [x] Run `mypy -p deepchem` and check no errors
  - [x] Run `flake8 <modified file> --count` and check no errors
  - [x] Run `python -m doctest <modified file>` and check no errors
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

**Comment by GreatRSingh** [comment]
Tasks:

- [ ] Port Layer
- [ ] Description in Documentation
- [ ] Example
- [ ] Shape and Properties Test
- [ ] PyTorch Comparision Test

**Review by shaipranesh2 — CHANGES_REQUESTED** [review]

File: deepchem/models/torch_models/layers.py [hunk 1]
```diff
@@ -2972,3 +2972,53 @@ def _dynamic_partition(self, input_tensor: torch.Tensor,
...
+        super(DTNNEmbedding, self).__init__(**kwargs)
+        self.n_embedding = n_embedding
+        self.periodic_table_length = periodic_table_length
+        self.init = init  # Set weight initialization
+
+        init_func: Callable = getattr(initializers, self.init)
+        self.embedding_list = init_func(
+            torch.empty([self.periodic_table_length, self.n_embedding]))
+
+    def __repr__(self) -> str:
```
[the remark this answers is not in the document]
Comment (GreatRSingh) [reply deepchem/models/torch_models/layers.py:3016]: I have wrote a new doc string comment. I am not sure if it is exactly what you suggested.
Can you give me an example.

**Comment by GreatRSingh** [comment]
@shaipranesh2 Ready For Review.

**Comment by shaipranesh2** [comment]
LGTM!
@rbharath Can you merge this PR.

File: deepchem/models/torch_models/layers.py [hunk 2]
```diff
@@ -2972,3 +2972,65 @@ def _dynamic_partition(self, input_tensor: torch.Tensor,
...
+    Assign initial atomic descriptors.
+
+    Parameters
+    ----------
+    n_embedding: int, optional
+        Number of features for each atom
+    periodic_table_length: int, optional
+        Length of embedding, 83=Bi
+    init: str, optional
+        Weight initialization for filters.
```
Comment (tonydavis629) [inline deepchem/models/torch_models/layers.py:2988]: Need to specify what options are possible

File: deepchem/models/torch_models/layers.py [hunk 3]
```diff
@@ -2972,3 +2972,65 @@ def _dynamic_partition(self, input_tensor: torch.Tensor,
         partitions = [input_tensor[mask] for mask in partition_masks]
 
         return partitions
+
+
+class DTNNEmbedding(nn.Module):
+    """DTNNEmbedding layer for DTNN model.
+    Assign initial atomic descriptors.
```
Comment (tonydavis629) [inline deepchem/models/torch_models/layers.py:2979]: Should include a reference to the original work describing DTNN

File: deepchem/models/tests/test_layers.py [hunk 4]
```diff
@@ -992,3 +992,25 @@ def test_set_gather():
...
+    ]
+    results_tf = [[0.64324486, -0.66274744, 0.2814387, 0.5478991, -0.32046735],
+                  [-0.26653743, 0.15180665, 0.21961051, -0.7263894, -0.4521287],
+                  [0.1925143, -0.5505201, -0.35381562, -0.7409675, 0.6427947]]
+    embedding_layer_torch = torch_layers.DTNNEmbedding(5, 5, 'xavier_uniform_')
+    embedding_layer_torch.embedding_list = torch.tensor(embeddings_tf)
+    result_torch = embedding_layer_torch(torch.tensor([3, 2, 4]))
+    assert torch.allclose(torch.tensor(results_tf), result_torch)
+    assert result_torch.shape == (3, 5)
+    assert embedding_layer_torch.embedding_list.shape == (5, 5)
```
Comment (tonydavis629) [inline deepchem/models/tests/test_layers.py:1016]: what is `embedding_layer_torch.embedding_list`? This attribute is not part of the API of the class. The L1016 test is not checking any functionality, it is only taking a specified tensor and asserting its shape.

File: deepchem/models/torch_models/layers.py [hunk 5]
```diff
@@ -2972,3 +2972,65 @@ def _dynamic_partition(self, input_tensor: torch.Tensor,
...
+            Weight initialization for filters.
+
+        """
+        return f'{self.__class__.__name__}(n_embedding={self.n_embedding}, periodic_table_length={self.periodic_table_length}, init={self.init})'
+
+    def forward(self, inputs):
+        """
+        parent layers: atom_number
+        """
+        atom_number = inputs
```
Comment (tonydavis629) [inline deepchem/models/torch_models/layers.py:3035]: Needs type hints. What is inputs?

File: deepchem/models/torch_models/layers.py (hunk 2 above)
Comment (GreatRSingh) [reply deepchem/models/torch_models/layers.py:2988]: I think 'init' name is apt there. Can you suggest what i should change it to?
I have written the options available for init. I have skipped some of the options as there seems to be too many of them, do you want that i write all of them.

File: deepchem/models/torch_models/layers.py (hunk 3 above)
Comment (GreatRSingh) [reply deepchem/models/torch_models/layers.py:2979]: Done

File: deepchem/models/tests/test_layers.py (hunk 4 above)
Comment (GreatRSingh) [reply deepchem/models/tests/test_layers.py:1016]: Removed.

File: deepchem/models/torch_models/layers.py (hunk 5 above)
Comment (GreatRSingh) [reply deepchem/models/torch_models/layers.py:3035]: I have made some changes here. I think something better can be done here do you have any suggestion.

File: deepchem/models/torch_models/layers.py [hunk 6]
```diff
@@ -2972,3 +2972,77 @@ def _dynamic_partition(self, input_tensor: torch.Tensor,
...
+        init: str, optional
+            Options: {xavier_uniform_, xavier_normal_, kaiming_uniform_, kaiming_normal_, trunc_normal_}
+            Weight initialization for filters.
+
+        """
+        return f'{self.__class__.__name__}(n_embedding={self.n_embedding}, periodic_table_length={self.periodic_table_length}, init={self.init})'
+
+    def forward(self, inputs: torch.Tensor):
+        """
+        parent layers: atom_number
```
Comment (tonydavis629) [inline deepchem/models/torch_models/layers.py:3040]: I'm not clear what this is supposed to mean?

Comment (GreatRSingh) [reply deepchem/models/torch_models/layers.py:3040]: It means that it is getting a tensor called 'atom_number' as input from the parent layers.
Here parent layers is the location where this layer is used.

Comment (GreatRSingh) [reply deepchem/models/torch_models/layers.py]: atom_number is used to choose the required embeddings.

**Comment by GreatRSingh** [comment]
@tonydavis629 @shaipranesh2 Ready for review.

File: deepchem/models/torch_models/layers.py [hunk 7]
```diff
@@ -2972,3 +2972,77 @@ def _dynamic_partition(self, input_tensor: torch.Tensor,
...
+
+    def forward(self, inputs: torch.Tensor):
+        """parent layers: atom_number
+
+        Parameters
+        ----------
+        inputs: torch.Tensor
+            Tensor containing the Atom number of Atoms whose embeddings are requested.
+
+        """
```
Comment (tonydavis629) [inline deepchem/models/torch_models/layers.py:3046]: more explanation on what the function actually does, remove 'parent layers' if it doesn't have an explanation. Then also requires a return value with its type, shape and explanation.

**Comment by tonydavis629** [comment]
You have a yapf error in layers.py

File: deepchem/models/torch_models/layers.py [hunk 8]
```diff
@@ -2972,3 +2972,83 @@ def _dynamic_partition(self, input_tensor: torch.Tensor,
         partitions = [input_tensor[mask] for mask in partition_masks]
 
         return partitions
+
+
+class DTNNEmbedding(nn.Module):
+    """DTNNEmbedding layer for DTNN model.
+    Assign initial atomic descriptors. [1]_
```
Comment (rbharath) [inline deepchem/models/torch_models/layers.py:2979]: Let's add a longer description here. Say about a small paragraph that explains what this layer does conceptually.

Comment (GreatRSingh) [reply deepchem/models/torch_models/layers.py:2979]: Done

**Review by tonydavis629 — APPROVED** [review]
Approved

**Review by shaipranesh2 — APPROVED** [review]

**Pull request merged**